// angular-cli-ghpages/ng-add.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const ng_add_1 = require("./ng-add");
const PROJECT_NAME = 'pie-ka-chu';
const PROJECT_ROOT = 'pirojok';
const OTHER_PROJECT_NAME = 'pi-catch-you';
describe('ng-add', () => {
    describe('generating files', () => {
        let tree;
        beforeEach(() => {
            tree = schematics_1.Tree.empty();
            tree.create('angular.json', JSON.stringify(generateAngularJson()));
        });
        it('generates new files if starting from scratch', () => __awaiter(void 0, void 0, void 0, function* () {
            const result = ng_add_1.ngAdd({
                project: PROJECT_NAME
            })(tree, {});
            expect(result.read('angular.json').toString()).toEqual(initialAngularJson);
        }));
        it('overrides existing files', () => __awaiter(void 0, void 0, void 0, function* () {
            const tempTree = ng_add_1.ngAdd({
                project: PROJECT_NAME
            })(tree, {});
            const result = ng_add_1.ngAdd({
                project: OTHER_PROJECT_NAME
            })(tempTree, {});
            const actual = result.read('angular.json').toString();
            expect(actual).toEqual(overwriteAngularJson);
        }));
    });
    describe('error handling', () => {
        it('fails if project not defined', () => {
            const tree = schematics_1.Tree.empty();
            const angularJSON = generateAngularJson();
            delete angularJSON.defaultProject;
            tree.create('angular.json', JSON.stringify(angularJSON));
            expect(() => ng_add_1.ngAdd({
                project: ''
            })(tree, {})).toThrowError('No Angular project selected and no default project in the workspace');
        });
        it('Should throw if angular.json not found', () => __awaiter(void 0, void 0, void 0, function* () {
            expect(() => ng_add_1.ngAdd({
                project: PROJECT_NAME
            })(schematics_1.Tree.empty(), {})).toThrowError('Could not find angular.json');
        }));
        it('Should throw if angular.json can not be parsed', () => __awaiter(void 0, void 0, void 0, function* () {
            const tree = schematics_1.Tree.empty();
            tree.create('angular.json', 'hi');
            expect(() => ng_add_1.ngAdd({
                project: PROJECT_NAME
            })(tree, {})).toThrowError('Could not parse angular.json');
        }));
        it('Should throw if specified project does not exist ', () => __awaiter(void 0, void 0, void 0, function* () {
            const tree = schematics_1.Tree.empty();
            tree.create('angular.json', JSON.stringify({ projects: {} }));
            expect(() => ng_add_1.ngAdd({
                project: PROJECT_NAME
            })(tree, {})).toThrowError('The specified Angular project is not defined in this workspace');
        }));
        it('Should throw if specified project is not application', () => __awaiter(void 0, void 0, void 0, function* () {
            const tree = schematics_1.Tree.empty();
            tree.create('angular.json', JSON.stringify({
                projects: { [PROJECT_NAME]: { projectType: 'pokemon' } }
            }));
            expect(() => ng_add_1.ngAdd({
                project: PROJECT_NAME
            })(tree, {})).toThrowError('Deploy requires an Angular project type of "application" in angular.json');
        }));
        it('Should throw if app does not have architect configured', () => __awaiter(void 0, void 0, void 0, function* () {
            const tree = schematics_1.Tree.empty();
            tree.create('angular.json', JSON.stringify({
                projects: { [PROJECT_NAME]: { projectType: 'application' } }
            }));
            expect(() => ng_add_1.ngAdd({
                project: PROJECT_NAME
            })(tree, {})).toThrowError('Cannot read the output path (architect.build.options.outputPath) of the Angular project "pie-ka-chu" in angular.json');
        }));
    });
});
function generateAngularJson() {
    return {
        defaultProject: PROJECT_NAME,
        projects: {
            [PROJECT_NAME]: {
                projectType: 'application',
                root: PROJECT_ROOT,
                architect: {
                    build: {
                        options: {
                            outputPath: 'dist/ikachu'
                        }
                    }
                }
            },
            [OTHER_PROJECT_NAME]: {
                projectType: 'application',
                root: PROJECT_ROOT,
                architect: {
                    build: {
                        options: {
                            outputPath: 'dist/ikachu'
                        }
                    }
                }
            }
        }
    };
}
const initialAngularJson = `{
  "defaultProject": "pie-ka-chu",
  "projects": {
    "pie-ka-chu": {
      "projectType": "application",
      "root": "pirojok",
      "architect": {
        "build": {
          "options": {
            "outputPath": "dist/ikachu"
          }
        },
        "deploy": {
          \"builder\": \"angular-cli-ghpages:deploy\",
          "options": {}
        }
      }
    },
    "pi-catch-you": {
      "projectType": "application",
      "root": "pirojok",
      "architect": {
        "build": {
          "options": {
            "outputPath": "dist/ikachu"
          }
        }
      }
    }
  }
}`;
const overwriteAngularJson = `{
  "defaultProject": "pie-ka-chu",
  "projects": {
    "pie-ka-chu": {
      "projectType": "application",
      "root": "pirojok",
      "architect": {
        "build": {
          "options": {
            "outputPath": "dist/ikachu"
          }
        },
        "deploy": {
          \"builder\": \"angular-cli-ghpages:deploy\",
          "options": {}
        }
      }
    },
    "pi-catch-you": {
      "projectType": "application",
      "root": "pirojok",
      "architect": {
        "build": {
          "options": {
            "outputPath": "dist/ikachu"
          }
        },
        "deploy": {
          "builder": "angular-cli-ghpages:deploy",
          "options": {}
        }
      }
    }
  }
}`;
