// angular-cli-ghpages/deploy/actions.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@angular-devkit/core");
const actions_1 = __importDefault(require("./actions"));
let context;
const mockEngine = { run: (_, __, __2) => Promise.resolve() };
const PROJECT = 'pirojok-project';
const BUILD_TARGET = {
    name: `${PROJECT}:build:production`
};
describe('Deploy Angular apps', () => {
    beforeEach(() => initMocks());
    it('should invoke the builder', () => __awaiter(void 0, void 0, void 0, function* () {
        const spy = spyOn(context, 'scheduleTarget').and.callThrough();
        yield actions_1.default(mockEngine, context, BUILD_TARGET, {});
        expect(spy).toHaveBeenCalledWith({
            target: 'build',
            configuration: 'production',
            project: PROJECT
        }, {});
    }));
    it('should invoke the builder with the baseHref', () => __awaiter(void 0, void 0, void 0, function* () {
        const spy = spyOn(context, 'scheduleTarget').and.callThrough();
        yield actions_1.default(mockEngine, context, BUILD_TARGET, { baseHref: '/folder' });
        expect(spy).toHaveBeenCalledWith({
            target: 'build',
            configuration: 'production',
            project: PROJECT
        }, { baseHref: '/folder' });
    }));
    it('should invoke engine.run', () => __awaiter(void 0, void 0, void 0, function* () {
        const spy = spyOn(mockEngine, 'run').and.callThrough();
        yield actions_1.default(mockEngine, context, BUILD_TARGET, {});
        expect(spy).toHaveBeenCalledWith('dist/some-folder', {}, context.logger);
    }));
    describe('error handling', () => {
        it('throws if there is no target project', () => __awaiter(void 0, void 0, void 0, function* () {
            context.target = undefined;
            try {
                yield actions_1.default(mockEngine, context, BUILD_TARGET, {});
                fail();
            }
            catch (e) {
                expect(e.message).toMatch(/Cannot execute the build target/);
            }
        }));
        it('throws if app building fails', () => __awaiter(void 0, void 0, void 0, function* () {
            context.scheduleTarget = (_, __, ___) => Promise.resolve({
                result: Promise.resolve(createBuilderOutputMock(false))
            });
            try {
                yield actions_1.default(mockEngine, context, BUILD_TARGET, {});
                fail();
            }
            catch (e) {
                expect(e.message).toEqual('Error while building the app.');
            }
        }));
    });
});
const initMocks = () => {
    context = {
        target: {
            configuration: 'production',
            project: PROJECT,
            target: 'foo'
        },
        builder: {
            builderName: 'mock',
            description: 'mock',
            optionSchema: false
        },
        currentDirectory: 'cwd',
        id: 1,
        logger: new core_1.logging.NullLogger(),
        workspaceRoot: 'cwd',
        addTeardown: _ => { },
        validateOptions: _ => Promise.resolve({}),
        getBuilderNameForTarget: () => Promise.resolve(''),
        analytics: null,
        getTargetOptions: (_) => Promise.resolve({
            outputPath: 'dist/some-folder'
        }),
        reportProgress: (_, __, ___) => { },
        reportStatus: (_) => { },
        reportRunning: () => { },
        scheduleBuilder: (_, __, ___) => Promise.resolve({}),
        scheduleTarget: (_, __, ___) => Promise.resolve({
            result: Promise.resolve(createBuilderOutputMock(true))
        })
    };
};
const createBuilderOutputMock = (success) => {
    return {
        info: { info: null },
        error: undefined,
        success: success,
        target: {}
    };
};
